### PR TITLE
QAP 505 : Adding  TestNG java schedule and manual workflow

### PR DESCRIPTION
**Pull request by ja3yntwillis**

### CODE REVIEW

File: workflow-templates/im-test-testng-schedule.yml [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+          restore-keys: |
+             ${{ runner.os }}-maven-
+
+      - name: Install Maven and Dependencies
+        # Installs Maven dependencies if the cache is not found.
+        if: steps.has-cache.outputs.cache-hit != 'true'
+        run: |
+          echo "Setting up Maven and installing dependencies"
+          chmod +x ${{ env.MAVEN_SETTINGS_PATH }}
+          ${{ env.MAVEN_SETTINGS_PATH }}
```
Comment (jsclifford) [inline workflow-templates/im-test-testng-schedule.yml:100]:
```suggestion
  maven-cache:
    # Sets up the Maven cache to speed up dependency installation
    runs-on: im-linux

    defaults:
      run:
        shell: bash

    outputs:
      MAVEN_CACHE_KEY: ${{ env.MAVEN_CACHE_KEY }}

    steps:
      - name: Checkout Repository
        uses: actions/checkout@v4
        with:
          ref: ${{ env.GITHUB_REF }}

      - name: Setup Java
        uses: actions/setup-java@v4
        if: steps.has-cache.outputs.cache-hit != 'true'
        with:
          distribution: 'temurin'
          java-version: ${{ env.JAVA_VERSION }}
          
      - name: Set up Maven Cache Keys
         # Generate a cache key based on the pom.xml file. If it changes, a new cache will be created.
         run: echo "MAVEN_CACHE_KEY=maven_repo-${{ hashFiles('**/pom.xml') }}" >> $GITHUB_ENV

      - name: Check for Maven Cache
         id: has-cache
         uses: actions/cache@v4
         with:
           lookup-only: true
           key: ${{ env.MAVEN_CACHE_KEY }}
           path: ${{ env.MAVEN_CACHE_PATH }} # Cache path for Maven dependencies.

      - name: Set up Maven
         if: steps.has-cache.outputs.cache-hit != 'true'
        # Steps to generate the shell script is available at  https://github.com/im-practices/techhub/blob/main/docs/Azure/content/Health-Checks/TestNG-UI-Health-Check.md Step no 6
        # This step will run the shell script to setup Maven with the modified settings.xml file.
        run: |
          chmod +x ${{ env.MAVEN_SETTINGS_PATH }}
          ${{ env.MAVEN_SETTINGS_PATH }}
        shell: bash

      - name: Build Project using Maven 
         if: steps.has-cache.outputs.cache-hit != 'true'
        run: ${{ env.MAVEN_BIN }} clean install

      - name: Install Maven and Dependencies
        # Installs Maven dependencies if the cache is not found.
        if: steps.has-cache.outputs.cache-hit != 'true'
        run: |
          echo "Setting up Maven and installing dependencies"
          chmod +x ${{ env.MAVEN_SETTINGS_PATH }}
          ${{ env.MAVEN_SETTINGS_PATH }}
```

File: workflow-templates/im-test-testng-schedule.yml [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        with:
+          distribution: 'temurin'
+          java-version: ${{ env.JAVA_VERSION }}
+
+      - name: Download Maven Cache
+        id: restore-maven-cache
+        # Restore the cached Maven dependencies if available.
+        uses: actions/cache/restore@v4
+        with:
+          key: ${{ needs.setup-cache.outputs.MAVEN_CACHE_KEY }}
```
Comment (jsclifford) [inline workflow-templates/im-test-testng-schedule.yml:136]:
```suggestion
          key: ${{ needs.maven-cache.outputs.MAVEN_CACHE_KEY }}
```

**Review by jsclifford — CHANGES_REQUESTED** [review]
Please update to maven-cache step and you need to skip steps when the cache is hit.

File: workflow-templates/im-test-testng-schedule.yml [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+          echo "Setting up Maven and installing dependencies"
+          chmod +x ${{ env.MAVEN_SETTINGS_PATH }}
+          ${{ env.MAVEN_SETTINGS_PATH }}
+        shell: bash
+
+      - name: Build Project using Maven 
+        if: steps.has-cache.outputs.cache-hit != 'true'
+        run: ${{ env.MAVEN_BIN }} clean install
+
+
```
Comment (jsclifford) [inline workflow-templates/im-test-testng-schedule.yml:95]: You can get rid of the extra lines.

File: workflow-templates/im-test-testng-schedule.yml [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+#
+# TODO Prerequisites:
+#    - Make sure you have the Javaagent and applicationinsights.json in resources folder.
+#    - Update your java selenium solution to export test results to Application Insights using the im.exporttoappinsights jar dependencies. 
+#    - Add secret APPLICATIONINSIGHTS_CONNECTION_STRING to every environment you are testing.
+#    - Modify your Maven profile fetch artifacts from github packages.
+#      Detailed setup guidance can be found here:  https://github.com/im-practices/techhub/blob/main/docs/Azure/content/Health-Checks/TestNG-UI-Health-Check.md
+#    - Create the shell script to setup Maven with the modified settings.xml file.details are at the following link Step 4,5
+#    - https://github.com/im-practices/techhub/blob/main/docs/Azure/content/Health-Checks/TestNG-UI-Health-Check.md
+name: 🧪 Run Java Selenium TestNG tests on Schedule
```
Comment (jsclifford) [inline workflow-templates/im-test-testng-schedule.yml:20]: Add a line between the comments and the name for better readability

File: workflow-templates/im-test-testng-schedule.yml [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+          chmod +x ${{ env.MAVEN_SETTINGS_PATH }}
+          ${{ env.MAVEN_SETTINGS_PATH }}
+        shell: bash
+
+      - name: Build Project using Maven 
+        if: steps.has-cache.outputs.cache-hit != 'true'
+        run: ${{ env.MAVEN_BIN }} clean install
+
+
+  testng-tests:
```
Comment (jsclifford) [inline workflow-templates/im-test-testng-schedule.yml:96]:
```suggestion
  selenium-maven-tests:
```
It is good practice to name the job after the type of test

**Review by jsclifford — COMMENTED** [review]
Looks good, but some minor suggestions.

File: workflow-templates/im-test-testng-schedule.yml (hunk 3 above)
Comment (ja3yntwillis) [reply workflow-templates/im-test-testng-schedule.yml:95]: removed  blank line

File: workflow-templates/im-test-testng-schedule.yml (hunk 4 above)
Comment (ja3yntwillis) [reply workflow-templates/im-test-testng-schedule.yml:20]: added line

**Comment by mike-schenk** [comment]
I don't think we should take on the burden of maintaining and troubleshooting Java environments on our runners when the same needs can be met with other environments. I.e. Selenium tests can be authored in TypeScript, JavaScript or C# which are all standard languages.

I vote to close the PR.

**Comment by ja3yntwillis** [comment]
@mike-schenk  

this workflow is added for the IM Customer Engagement (PEGA) team as they are currently using java based automation framework, this workflow should only be used if java is required by external vendor testing. 

cc @jcfrperu @Oyun1 @jsclifford  @jacobjithin

**Comment by mike-schenk** [comment]
> @mike-schenk
> 
> this workflow is added for the IM Customer Engagement (PEGA) team as they are currently using java based automation framework, this workflow should only be used if java is required by external vendor testing.
> 
> cc @jcfrperu @Oyun1 @jsclifford @jacobjithin

Since no other teams should use this, it shouldn't be in our shared templates. The Pega team can keep their copy in their own repository though.

**Pull request closed without merging**